lib/mongoid/contextual/mongo.rb

```diff
@@ -1038,6 +1043,24 @@ def process_raw_docs(raw_docs, limit)
         limit ? docs : docs.first
       end
 
+      # Queries whether the current context is valid for use with
+      # the #count_documents? predicate. A context is valid if it
+      # does not include a `$where` operator.
+      #
+      # @return [ true | false ] whether or not the current context
+      #   excludes a `$where` operator.
+      def valid_for_count_documents?(hash = view.filter)
+        # Note that `view.filter` is a BSON::Document, and all keys in a
+        # BSON::Document are strings; we don't need to worry about symbol
+        # representations of `$where`.
+        hash.keys.each do |key|
+          return false if key == '$where'
+          return false if hash[key].is_a?(Hash) && !valid_for_count_documents(hash[key])
```

Comment:
```suggestion
          return false if hash[key].is_a?(Hash) && !valid_for_count_documents?(hash[key])
```

I think there's a typo there